File: frontend/src/api/summaryAdjustment.ts
import { del, get, post, put } from "./index";

export type SummaryAdjustmentMetric = "gmv" | "return_cost";
export type SummaryAdjustmentDirection = "increase" | "decrease";

export interface SummaryAdjustment {
    id: number;
    org_id: number;
    source_year: number;
    source_month: number;
    platform_name: string;
    shop_id: number;
    shop_name: string;
    metric_key: SummaryAdjustmentMetric;
    metric_label: string;
    direction: SummaryAdjustmentDirection;
    amount: number;
    adjustment_amount: number;
    remark?: string | null;
    created_by: number;
    updated_by?: number | null;
    created_at: string;
    updated_at: string;
}

export interface SummaryAdjustmentListParams {
    source_year: number;
    source_month: number;
    platform_name: string;
    shop_id: number;
    metric_key?: SummaryAdjustmentMetric;
}

export interface SummaryAdjustmentPayload {
    source_year: number;
    source_month: number;
    platform_name: string;
    shop_id: number;
    shop_name: string;
    metric_key: SummaryAdjustmentMetric;
    direction: SummaryAdjustmentDirection;
    amount: number;
    remark?: string;
}

export interface SummaryAdjustmentUpdatePayload {
    metric_key?: SummaryAdjustmentMetric;
    direction?: SummaryAdjustmentDirection;
    amount?: number;
    remark?: string;
}

export function getSummaryAdjustments(params: SummaryAdjustmentListParams) {
    return get<SummaryAdjustment[]>("/summary-adjustments", params);
}

export function createSummaryAdjustment(data: SummaryAdjustmentPayload) {
    return post<SummaryAdjustment>("/summary-adjustments", data);
}

export function updateSummaryAdjustment(id: number, data: SummaryAdjustmentUpdatePayload) {
    return put<SummaryAdjustment>(`/summary-adjustments/${id}`, data);
}

export function deleteSummaryAdjustment(id: number) {
    return del<null>(`/summary-adjustments/${id}`);
}

File: backend/app/schemas/summary_adjustment.py
from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field, field_validator


SummaryAdjustmentMetric = Literal["gmv", "return_cost"]
SummaryAdjustmentDirection = Literal["increase", "decrease"]

SUMMARY_ADJUSTMENT_METRIC_LABELS = {
    "gmv": "实收GMV",
    "return_cost": "退货费用及其他费用",
}


class SummaryAdjustmentCreate(BaseModel):
    source_year: int = Field(..., ge=2000, le=2100, description="数据表上传年份")
    source_month: int = Field(..., ge=1, le=12, description="数据表上传月份")
    platform_name: str = Field(..., min_length=1, max_length=50, description="平台编码")
    shop_id: int = Field(..., description="店铺ID")
    shop_name: str = Field(..., min_length=1, max_length=200, description="店铺名称")
    metric_key: SummaryAdjustmentMetric = Field(..., description="调整指标")
    direction: SummaryAdjustmentDirection = Field(..., description="调整方向")
    amount: Decimal = Field(..., gt=0, decimal_places=2, description="调整金额绝对值")
    remark: str | None = Field(None, max_length=2000, description="备注")

    @field_validator("platform_name", "shop_name")
    @classmethod
    def strip_required_text(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("不能为空")
        return stripped

    @field_validator("remark")
    @classmethod
    def strip_optional_text(cls, value: str | None) -> str | None:
        return value.strip() if value else None


class SummaryAdjustmentUpdate(BaseModel):
    metric_key: SummaryAdjustmentMetric | None = Field(None, description="调整指标")
    direction: SummaryAdjustmentDirection | None = Field(None, description="调整方向")
    amount: Decimal | None = Field(None, gt=0, decimal_places=2, description="调整金额绝对值")
    remark: str | None = Field(None, max_length=2000, description="备注")

    @field_validator("remark")
    @classmethod
    def strip_optional_text(cls, value: str | None) -> str | None:
        return value.strip() if value else None


class SummaryAdjustmentOut(BaseModel):
    id: int
    org_id: int
    source_year: int
    source_month: int
    platform_name: str
    shop_id: int
    shop_name: str
    metric_key: str
    metric_label: str
    direction: str
    amount: float
    adjustment_amount: float
    remark: str | None = None
    created_by: int
    updated_by: int | None = None
    created_at: datetime
    updated_at: datetime


def signed_adjustment_amount(direction: str, amount: Decimal) -> Decimal:
    return amount if direction == "increase" else -amount
